src/tests/reactA11yAccessibleHeadingsRuleTests.ts

```diff
@@ -0,0 +1,211 @@
+                "name": Utils.absolutePath("file.tsx"),
+                "ruleName": "react-a11y-accessible-headings",
+                "startPosition": { "character": 30, "line": 5 }
+            }
+        ]);
+    });
+
+    it('should fail on self-closing heading', () : void => {
+        const script : string = `
+            import React = require('react');
+            const heading = <h1/>;
+        `;
+
+        TestHelper.assertViolations(ruleName, script, [
+            {
+                "failure": "Heading elements must not be empty",
+                "name": Utils.absolutePath("file.tsx"),
+                "ruleName": "react-a11y-accessible-headings",
+                "startPosition": { "character": 29, "line": 3 }
+            }
+        ]);
+    });
+    it('should pass when headings are ordered correctly', () : void => {
+        const script : string = `
+            import React = require('react');
+            const heading1 = <h1>test</h1>;
+            const heading3 = <h2>test</h2>;
+            const heading4 = <h3>test</h3>;
+            const heading5 = <h3>test</h3>;
+            const heading6 = <h4>test</h4>;
+            const heading7 = <h5>test</h5>;
+            const heading8 = <h6>test</h6>;
+            const heading9 = <h6>test</h6>;
+        `;
+
+        TestHelper.assertViolations(ruleName, script, []);
+    });
+    it('should fail when headings are ordered incorrectly', () : void => {
+        const script : string = `
+            import React = require('react');
+            const heading1 = <h3>test</h3>;
+            const heading2 = <h2>test</h2>;
+        `;
+
+        TestHelper.assertViolations(ruleName, script, [
+            {
+                "failure": "Heading elements should be used for structuring information on the page",
+                "name": Utils.absolutePath("file.tsx"),
+                "ruleName": "react-a11y-accessible-headings",
+                "startPosition": { "character": 30, "line": 4 }
+            }
+        ]);
+    });
+    it('should fail when heading numbers are not increasing by one at a time', () : void => {
+        const script : string = `
+            import React = require('react');
+            const heading1 = <h3>test</h3>;
+            const heading2 = <h5>test</h5>;
+        `;
+
+        TestHelper.assertViolations(ruleName, script, [
+            {
+                "failure": "Heading elements should be used for structuring information on the page",
+                "name": Utils.absolutePath("file.tsx"),
+                "ruleName": "react-a11y-accessible-headings",
+                "startPosition": { "character": 30, "line": 4 }
+            }
+        ]);
+    });
+    it('should pass when H1 heading text is a substring of title text', () : void => {
+        const script : string = `
+            import React = require('react');
+            const title = <title>"I'm so great look at me..."</title>;
```

Comment:
> `<title>...</title>`
.
It looks like you're keeping track of the last known title text with `captureTitleText` / `capturedTitleText`, then comparing later nodes against it. That's making two assumptions in this rule that doesn't always hold true:

* The file that contains headings will also contain the `<title>`
* Files contain nodes in order

You can't actually set the `<title>` node easily in JS apps since it's part of the `<head`>. See https://github.com/gaearon/react-document-title

I don't think there's a way around these limitations. You'll likely have to get rid of this logic ☹.

Reply:
Similarly, the rule's tests should make sure that it doesn't flag files where multiple components are able to create an `<h1>` element, as it might be that only one is ever called. This should not cause a complaint:

```typescript
export const createTitleA = () => <h1>A</h1>;
export const createTitleB = () => <h1>B</h1>;
```

Reply:
Thanks for this review @JoshuaKGoldberg !

Reply:
@JoshuaKGoldberg 

I think I completely missed a bunch of stuff after reading what you're describing in your last comment. This is relevant to issues 3 and 5.
I think I need to put those tests into context and re-implement a bunch of stuff. (like distinction between assignments and actual script...)

Reply:
Hi @JoshuaKGoldberg.
fixed all the issues, and re implemented some rule's logic. (I hope not too naively )